### PR TITLE
Repositories - import functionality

### PR DESCRIPTION
**Pull request by ZmagoD**
https://biosistemika.atlassian.net/browse/SCI-1336

### CODE REVIEW

File: app/models/repository.rb [hunk 1]
```diff
@@ -41,4 +63,90 @@ def copy(created_by, name)
...
+        custom_fields << nil
+        name_index = i
+      else
+        cf = repository_columns.find_by_id(v)
+        custom_fields << cf
+      end
+    end
+
+    # Now we can iterate through sample data and save stuff into db
+    (2..sheet.last_row).each do |i|
```
Comment (ZmagoD) [inline app/models/repository.rb]: :banana:

Comment (mz3944) [inline app/models/repository.rb]: :banana:

File: app/models/repository.rb [hunk 2]
```diff
@@ -15,6 +15,28 @@ class Repository < ActiveRecord::Base
...
+    if file.class == Paperclip::Attachment && file.is_stored_on_s3?
+      fa = file.fetch
+      file_path = fa.path
+    end
+    generate_file(filename, file_path)
+  end
+
+  def available_repository_fields
+    fields = {}
+    # First and foremost add sample name
```
Comment (mz3944) [inline app/models/repository.rb]: record name*

File: app/models/repository.rb [hunk 3]
```diff
@@ -41,4 +63,86 @@ def copy(created_by, name)
...
+        # Fill blank space, so our indices stay the same
+        custom_fields << nil
+        name_index = index
+      else
+        cf = repository_columns.find_by_id(value)
+        custom_fields << cf
+      end
+    end
+
+    # Now we can iterate through sample data and save stuff into db
```
Comment (mz3944) [inline app/models/repository.rb]: record*

File: app/models/repository.rb [hunk 4]
```diff
@@ -41,4 +63,86 @@ def copy(created_by, name)
...
+
+  private
+
+  # def generate_new_columns(header)
+  #   rep_columns_names = self.repository_columns.pluck(:name).push('Name')
+  #   header.each do |cname|
+  #     next if rep_columns_names.include? cname
+  #     RepositoryColumn.create(repository: self, name: cname, data_type: 0)
+  #   end
+  # end
```
Comment (mz3944) [inline app/models/repository.rb]: Remove.

File: app/controllers/repositories_controller.rb [hunk 5]
```diff
@@ -180,6 +182,66 @@ def repository_table_index
...
+      unless params[:file]
+        repository_response(t('teams.parse_sheet.errors.no_file_selected'))
+        return
+      end
+      begin
+        if parsed_file.too_large?
+          repository_response(t('general.file.size_exceeded',
+                                file_size: Constants::FILE_MAX_SIZE_MB))
+        else
+          flash[:notice] = t('teams.parse_sheet.errors.empty_file')
```
Comment (mz3944) [inline app/controllers/repositories_controller.rb]: Put in if.

File: app/assets/javascripts/repositories/import/records_importer.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+          el.children().eq(currIndex).attr('disabled', 'disabled');
+        }
+
+        el.children().eq(previousIndex).removeAttr('disabled');
+      });
+
+      previousIndex = currIndex;
+    });
+
+    //  Create import samples ajax
```
Comment (mz3944) [inline app/assets/javascripts/repositories/import/records_importer.js]: Records.

File: app/assets/javascripts/repositories/import/records_importer.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    $('form#form-import').submit(function(e) {
+      // Check if we already uploading samples
```
Comment (mz3944) [inline app/assets/javascripts/repositories/import/records_importer.js]: records*.

File: app/assets/javascripts/repositories/import/records_importer.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      if (loadingSamples) {
```
Comment (mz3944) [inline app/assets/javascripts/repositories/import/records_importer.js]: loadingRecords*

File: app/assets/javascripts/repositories/import/records_importer.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        return false;
+      }
+      disabledOptions = $("option[disabled='disabled']");
+      disabledOptions.removeAttr('disabled');
+      loadingSamples = true;
+      animateSpinner();
+    }).on('ajax:success', function(ev, data, status) {
+      // Simply reload page to show flash and updated samples list
```
Comment (mz3944) [inline app/assets/javascripts/repositories/import/records_importer.js]: records list*

File: app/controllers/repositories_controller.rb [hunk 10]
```diff
@@ -199,9 +199,10 @@ def parse_sheet
         if parsed_file.too_large?
           repository_response(t('general.file.size_exceeded',
                                 file_size: Constants::FILE_MAX_SIZE_MB))
-        else
+        elsif parsed_file.empty?
           flash[:notice] = t('teams.parse_sheet.errors.empty_file')
-          redirect_to back and return if parsed_file.empty?
+          redirect_to back and return
```
Comment (mz3944) [inline app/controllers/repositories_controller.rb]: `and return` not needed, there's no more code after the closing tag.

**Pull request merged**